keps/sig-instrumentation/distributed-tracing-kep.md

```diff
@@ -0,0 +1,134 @@
+
+#### Instrumenting various Kubernetes components with pod startup traces
+
+As we move forward with this KEP, we hope to leverage the aforementioned trace utility package to trace basic pod-related operations across the scheduler and kubelet. A tentative implementation of this instrumentation is [located here](https://github.com/Monkeyanator/kubernetes/pull/15).
+
+In OpenCensus' Go implementation, span context is passed down through Go context. This will necessitate the threading of context across more of the Kubernetes codebase, which is a [desired outcome regardless](https://github.com/kubernetes/kubernetes/issues/815).
+
+Following these intial pod-related trace additions, trace instrumentation should be added in an ad-hoc fashion to the various Kubernetes components.
+
+
+### Out-of-tree changes
+
+#### Mutating admission webhook / trace context annotation
+
+For spans to be correlated as part of the same trace, we must generate a `span context`, serialize it, and embed it in target traced objects. To accomplish this, we have introduced an [out-of-tree mutating admission webhook](https://github.com/Monkeyanator/mutating-trace-admission-controller/tree/review).
+
+This mutating admission webhook generates a `span context`, which is the base64 encoded version of [this wire format](https://github.com/census-instrumentation/opencensus-specs/blob/master/encodings/BinaryEncoding.md#trace-context), and embeds it into the `trace.kubernetes.io/context` object annotation. The webhook can be configured to inject context into only target object types.
+
+The decision on whether or not to sample traces from a given object is made in this webhook, upon the generation of the trace context. In addition to this, the aforementioned OpenCensus agent offers ex-post-facto trace sampling.
+
+The proposed in-tree changes will utilize the span context annotation injected into objects with this webhook.
+
+#### Tracing best-practices documentation
+
+This KEP introduces a new form of instrumentation to Kubernetes, which necessiates the creation of guidelines for adding effective, standardized traces to Kubernetes components, [similar to what is found here for metrics](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md).
+
+This documentation should put forward standards for: 
+
+* How to name spans, attributes, and annotations
+* What kind of processes should be wrapped in a span
+* When to link spans to other traces
+* What kind of data makes for useful attributes
+
+Having these standards in place will ensure that our tracing instrumentation works well with all backends, and that reviewers have concrete criteria to cross-check PRs against. 
+
+
+### Risks and Mitigations
```

Comment:
There aren't any risks we need to care about given that this is an alpha feature, and is disabled by default.  These all need solutions before beta.